src/main/java/jmgomez/apipolicy/config/RequestInterceptor.java

```diff
@@ -0,0 +1,62 @@
+package jmgomez.apipolicy.config;
+
+import jakarta.servlet.ServletException;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import jmgomez.apipolicy.service.PolicyService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.stereotype.Component;
+import org.springframework.web.method.HandlerMethod;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.HandlerMapping;
+
+import java.io.IOException;
+import java.util.Map;
+
+@Component
+public class RequestInterceptor implements HandlerInterceptor {
+
+    private final PolicyService policyService;
+
+    @Autowired
+    public RequestInterceptor(@Lazy PolicyService policyService) {
+        this.policyService = policyService;
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws ServletException, IOException {
+        String authUserId = SecurityContextHolder.getContext().getAuthentication().getName();
+
+        Map<String, String> variables = (Map<String, String>) request.getAttribute(HandlerMapping.URI_TEMPLATE_VARIABLES_ATTRIBUTE);
+
+        String userId = authUserId;
+        String policyId = variables.get("policyId");
+        String accidentId = variables.get("accidentId");
+
+        if(handler instanceof HandlerMethod) {
+            if(request.getRequestURI().equals("/policies") && userId != null) {
+              return true;
+            } else if (request.getRequestURI().equals("/policies/" + policyId) && policyId != null && isPolicyOfUser(policyId, userId)) {
+                return true;
+            } else if (request.getRequestURI().equals("/policies/" + policyId + "/accidents") && policyId != null && isPolicyOfUser(policyId, userId)) {
+                return true;
+            } else if (request.getRequestURI().equals("/policies/" + policyId + "/accidents/" + accidentId) && policyId != null && accidentId != null && isPolicyOfUser(policyId, userId) && isAccidentOfPolicy(accidentId, policyId)) {
+                return true;
+            }
+
+        }
+        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        response.getWriter().write("You have no access with this user");
```

Comment:
Is this necessary? You just broke your API as it's not congruent. Errors handled by the controller advice will return a JSON object, this one just writes a status and a string... not a well thought decision.

![image](https://github.com/user-attachments/assets/47d76497-78c6-46ca-9be5-7a664ff24ae1)